src/java.desktop/windows/classes/sun/awt/windows/ThemeReader.java

```diff
@@ -117,7 +117,11 @@ private static Long openThemeImpl(String widget, int dpi) {
 
     // this should be called only with writeLock held
     private static Long getThemeImpl(String widget, int dpi) {
-       return dpiAwareWidgetToTheme.computeIfAbsent(dpi, key -> new HashMap<>())
+        if (openThemeImpl(widget, dpi) == null || openThemeImpl(widget, dpi) == 0) {
```

Comment:
I think it would be cleaner and better if this is handled in `openThemeImpl()` itself. We can put a check for NULL `theme `being returned and return the one based on `defaultDPI`, if that is the case.

Reply:
I agree. The current approach calls `openThemeImpl` three times: once or twice with the passed `dpi` and then once with `defaultDPI` as the fallback. Moreover, it *leaks theme handles* if `openThemeImpl` returns non-zero value and it does so repeatedly even when there's a cached theme handle.

Handling the failure inside `openThemeImpl` seems a better option. Even though it results in a higher memory usage.

I believe `openTheme` returns 0 because the DPI of a printer is higher than that of screens, and Windows doesn't support such DPIs. The problem is likely reproducible when you print other components such as buttons.

Reply:
> Yes, `openTheme()` actually returns 0. …

This is why it can't be `null`.